### PR TITLE
Get report metadata from Twin in TRC

### PR DESCRIPTION
**Pull request by philipktlin**
update to get report metadata from Twin for TRC module; add unit tests.

### CODE REVIEW

File: test/modules/Modules.Test/TestResultCoordinator/TestUtilTest.cs [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+        [Fact]
+        public async Task TestGenerateTestResultReportsAsync_NullReportMetadata()
+        {
+            var mockLogger = new Mock<ILogger>();
+            var mockTestReportGeneratorFactory = new Mock<ITestReportGeneratorFactory>();
+
+            ArgumentNullException ex = await Assert.ThrowsAsync<ArgumentNullException>(() =>
+                 TestUtil.GenerateTestResultReportsAsync(
+                     "fakeTrackngId",
```
Comment (dylanbronson) [inline test/modules/Modules.Test/TestResultCoordinator/TestUtilTest.cs]: nit: typo

File: test/modules/TestResultCoordinator/Settings.cs [hunk 2]
```diff
@@ -131,51 +134,34 @@ public override string ToString()
             return $"Settings:{Environment.NewLine}{string.Join(Environment.NewLine, fields.Select(f => $"{f.Key}={f.Value}"))}";
         }
 
-        internal List<ITestReportMetadata> GetReportMetadataList()
+        internal async Task<List<ITestReportMetadata>> GetReportMetadataListAsync(ILogger logger)
```
Comment (ancaantochi) [inline test/modules/TestResultCoordinator/Settings.cs]: It looks like this method should be in the TestUtil. Same for GetResultSources

File: test/modules/TestResultCoordinator/TestUtil.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json.Linq;
+    using TestResultCoordinator.Reports;
+    using TestResultCoordinator.Reports.DirectMethod;
+
+    static class TestUtil
```
Comment (ancaantochi) [inline test/modules/TestResultCoordinator/TestUtil.cs]: Suggestion: rename to Util or ReportGeneratorUtil

File: test/modules/TestResultCoordinator/TestUtil.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            }
+        }
+
+        internal static List<ITestReportMetadata> ParseReportMetadataJson(string reportMetadataListJson, ILogger logger)
+        {
+            var reportMetadataList = new List<ITestReportMetadata>();
+
+            try
+            {
+                JObject reportMetadatas = JObject.Parse(reportMetadataListJson);
```
Comment (ancaantochi) [inline test/modules/TestResultCoordinator/TestUtil.cs]: Would deserializing the json to its corresponding report metadata be easier?

File: test/modules/TestResultCoordinator/Settings.cs (hunk 2 above)
Comment (philipktlin) [reply test/modules/TestResultCoordinator/Settings.cs]: This method used a class field to persist the result from GetTwin and ReportMetadata is part of settings.  Therefore I think it is better to be in the class compared to TestUtil static class.

File: test/modules/TestResultCoordinator/TestUtil.cs (hunk 3 above)
Comment (philipktlin) [reply test/modules/TestResultCoordinator/TestUtil.cs]: It is not only used for report generation; how about TestReportUtil.

File: test/modules/TestResultCoordinator/TestUtil.cs (hunk 4 above)
Comment (philipktlin) [reply test/modules/TestResultCoordinator/TestUtil.cs]: updated to use json deserializaiton

File: test/modules/TestResultCoordinator/Settings.cs (hunk 2 above)
Comment (varunpuranik) [inline test/modules/TestResultCoordinator/Settings.cs]: With the "Current" pattern, it feels like the Twin should be obtained and cached in Create() itself.

File: test/modules/TestResultCoordinator/TestReportUtil.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            catch (Exception ex)
+            {
+                logger.LogError(ex, $"Incorrect parsing for report metadata list: {reportMetadataListJson}{Environment.NewLine}");
+                throw;
+            }
+
+            return reportMetadataList;
+        }
+
+        private static TEnum GetEnumValueFromReportMetadata<TEnum>(JToken metadata, string key)
```
Comment (varunpuranik) [inline test/modules/TestResultCoordinator/TestReportUtil.cs]: Nit: avoid private access modifier on methods and fields

File: test/modules/TestResultCoordinator/Settings.cs (hunk 2 above)
Comment (philipktlin) [reply test/modules/TestResultCoordinator/Settings.cs]: Current pattern is the singleton pattern to ensure only 1 Settings in the process.  I think it is reasonable for this class to provide methods to manage more setting-related data.

If you prefer to change Create method become async, then I need to:
1. Rename Create method to GetOrCreateAsync.
2. Update all usage of Settings.Current to become Settings.GetOrCreateAsync; which will use a simple lock and check whether to return cached object or create a new instance.

Please confirm which implementation you recommended; or suggest another way.  thanks.

**Pull request merged**